### PR TITLE
DirectX offscreen rendering for Swing Graphics2D

### PR DESCRIPTION
**Pull request by elijah-semyonov**

### CODE REVIEW

File: samples/SkiaAwtSample/src/main/kotlin/SkiaAwtSample/ClocksAwt.kt [hunk 1]
```diff
@@ -40,8 +40,8 @@ open class ClocksAwt(private val scaleProvider: () -> Float) : SkikoView {
           strokeWidth = 1f
         }
         val watchFillHover = Paint().apply { color = 0xFFE4FF01.toInt() }
-        for (x in 0 .. (width - 50) step 50) {
-            for (y in 20 .. (height - 50) step 50) {
+        for (x in 0 .. (width - 50) step 500) {
```
Comment (Walingar) [inline samples/SkiaAwtSample/src/main/kotlin/SkiaAwtSample/ClocksAwt.kt:43]: Let's revert these changes

File: samples/SkiaAwtSample/src/main/kotlin/SkiaAwtSample/ResizableSwingSkia.kt [hunk 2]
```diff
@@ -11,6 +11,7 @@ fun swingSkiaResize() = SwingUtilities.invokeLater {
     window.defaultCloseOperation = WindowConstants.DISPOSE_ON_CLOSE
     window.title = "Swing Window with Compose"
 
+//    var skiaPanel: SkiaPanel? = null
```
Comment (Walingar) [inline samples/SkiaAwtSample/src/main/kotlin/SkiaAwtSample/ResizableSwingSkia.kt:14]: Let's revert these changes as well

File: skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc [hunk 3]
```diff
@@ -0,0 +1,400 @@
+#ifdef SK_DIRECT3D
```
Comment (Walingar) [inline skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc:1]: @igordmn @MatkovIvan please take a look at this code

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/swing/Direct3DSwingRedrawer.kt [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    override fun onRender(g: Graphics2D, width: Int, height: Int, nanoTime: Long) {
+        val alignedWidth = if (width % widthSizeAlignment != 0) {
+            width + widthSizeAlignment - (width % widthSizeAlignment);
+        } else {
+            width
+        }
+
+        val newTexturePtr = getRenderTargetTexture(device, alignedWidth, height)
+
+        if (newTexturePtr == 0.toLong()) {
```
Comment (Walingar) [inline skiko/src/awtMain/kotlin/org/jetbrains/skiko/swing/Direct3DSwingRedrawer.kt:60]: Minor: 0.toLong() -> 0L

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/swing/Direct3DSwingRedrawer.kt [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            throw RenderException("Can't allocate DirectX resources")
+        }
+
+        if (newTexturePtr != texturePtr) {
+            texturePtr = newTexturePtr
```
Comment (Walingar) [inline skiko/src/awtMain/kotlin/org/jetbrains/skiko/swing/Direct3DSwingRedrawer.kt:65]: We should close renderTarget and surface in this case.
BTW can we create surface and renderTarget on each render as we do with Metal? Does it significantly affect performance?

**Review by Walingar — CHANGES_REQUESTED** [review]

File: skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+        d3dDevice->backendContext.fQueue = queue;
+        d3dDevice->backendContext.fProtectedContext = GrProtected::kNo;
+
+        auto& textureDesc = d3dDevice->textureDesc;
+        textureDesc.Dimension = D3D12_RESOURCE_DIMENSION_TEXTURE2D;
+        textureDesc.Alignment = 0;
+        textureDesc.Width = 0;
+        textureDesc.Height = 0;
+        textureDesc.DepthOrArraySize = 1;
+        textureDesc.MipLevels = 1;
```
Comment (MatkovIvan) [inline skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc:212]: I don't think that we need it

File: skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+        textureDesc.Layout = D3D12_TEXTURE_LAYOUT_UNKNOWN;
+        textureDesc.Flags = D3D12_RESOURCE_FLAG_ALLOW_RENDER_TARGET;
+
+        auto& readbackBufferDesc = d3dDevice->readbackBufferDesc;
+        readbackBufferDesc.Dimension = D3D12_RESOURCE_DIMENSION_BUFFER;
+        readbackBufferDesc.Alignment = 0;
+        readbackBufferDesc.Width = 0;
+        readbackBufferDesc.Height = 1;
+        readbackBufferDesc.DepthOrArraySize = 1;
+        readbackBufferDesc.MipLevels = 1;
```
Comment (MatkovIvan) [inline skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc:225]: Same

File: skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+        }
+
+        memcpy(bytesPtr, readbackBufferBytesPtr, rangeLength);
+
+        D3D12_RANGE emptyRange{ 0, 0 };
+        device->readbackBuffer->Unmap
+        (
+            0,
+            &emptyRange
+        );
```
Comment (MatkovIvan) [inline skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc:380]:
```suggestion
        device->readbackBuffer->Unmap(0, &emptyRange);
```

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/swing/Direct3DSwingRedrawer.kt [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+            val format = 28 // DXGI_FORMAT_R8G8B8A8_UNORM
```
Comment (MatkovIvan) [inline skiko/src/awtMain/kotlin/org/jetbrains/skiko/swing/Direct3DSwingRedrawer.kt:67]: Add a const for that

File: samples/SkiaAwtSample/src/main/kotlin/SkiaAwtSample/ResizableSwingSkia.kt [hunk 10]
```diff
@@ -22,6 +23,10 @@ fun swingSkiaResize() = SwingUtilities.invokeLater {
         }
     }
 
+//    skiaPanel = SkiaPanel().also {
```
Comment (igordmn) [inline samples/SkiaAwtSample/src/main/kotlin/SkiaAwtSample/ResizableSwingSkia.kt:26]: Checked it on my machine, FPS of SkiaSwingPanel is the same as with SkiaPanel 🎉

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/swing/Direct3DSwingRedrawer.kt (hunk 5 above)
Comment (elijah-semyonov) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/swing/Direct3DSwingRedrawer.kt:65]: Yes, indeed, they leak now on each resize.  Recreating those makes sense when their wrapping entity or usage options change, which is not the case. I don't think it affects performance since those are lightweight logical entities (assuming Skia doesn't do something dumb or too smart because of this), but why would we opt-in into that way?

File: skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc [hunk 11]
```diff
@@ -0,0 +1,400 @@
...
+            return 0;
+        }
+        if (adapterPtr == 0) {
+            return 0;
+        }
+        gr_cp<IDXGIAdapter1> adapter((IDXGIAdapter1 *) adapterPtr);
+
+        D3D_FEATURE_LEVEL maxSupportedFeatureLevel = D3D_FEATURE_LEVEL_12_0;
+        D3D_FEATURE_LEVEL featureLevels[] = {
+            // TODO add D3D_FEATURE_LEVEL_12_2
```
[the remark this answers is not in the document]
Comment (elijah-semyonov) [reply skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc:147]: @Walingar do you remember why this line was included in the scratch version?

File: skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc [hunk 12]
```diff
@@ -0,0 +1,400 @@
...
+#include "../common/interop.hh"
+
+#include "d3d/GrD3DTypes.h"
+#include <d3d12sdklayers.h>
+#include "d3d/GrD3DBackendContext.h"
+#include <d3d12.h>
+#include <dxgi1_4.h>
+#include <dxgi1_6.h>
+
+class DirectXOffscreenDevice
```
[the remark this answers is not in the document]
Comment (elijah-semyonov) [reply skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc:23]: does it matter?

File: skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc (hunk 6 above)
Comment (elijah-semyonov) [reply skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc:212]: I'd better keep it in explicit initial state to display that they are set somewhere else.

Comment (elijah-semyonov) [reply skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc:212]: It is needed 

https://learn.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_resource_desc

File: skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc (hunk 7 above)
Comment (elijah-semyonov) [reply skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc:225]: > Height, DepthOrArraySize, and MipLevels must be 1.

File: skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (MatkovIvan) [reply skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc:23]: In case of class I'd not put everything to public and add methods to work with them (encapsulation, right?)

[the remark this answers is not in the document]
Comment (elijah-semyonov) [reply skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc:23]: You imply semantics which aren't declared somewhere in the project. Moreover, on one of previous PRs I was asked not to put any logic outside of JNI methods, because it makes it harder to navigate between JNI methods bodies and code (without a lot of tinkering).

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/swing/Direct3DSwingRedrawer.kt (hunk 9 above)
Comment (elijah-semyonov) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/swing/Direct3DSwingRedrawer.kt:67]: I think it's even better to obtain it via JNI call and avoid magic numbers

File: skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (elijah-semyonov) [reply skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc:23]: Writing this code properly (using smart pointers, encapsulating all within methods) would turn this into code I was specifically asked not to produce.

File: skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (Walingar) [reply skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc:147]: I don't remember :) We can remove it I think

**Review by Walingar — APPROVED** [review]

**Comment by Walingar** [comment]
@elijah-semyonov I've refactored the code, can you please take a look at the changes?

File: skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc [hunk 13]
```diff
@@ -0,0 +1,424 @@
...
+        }
+
+        if (readbackBufferResource) {
+            readbackBufferResource->Release();
+        }
+    }
+
+    int readbackBufferWidth() {
+        return readbackBufferWidth(width, height);
+    }
```
Comment (elijah-semyonov) [inline skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc:131]: Isn't it redundant?

Comment (Walingar) [reply skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc:131]: It is used in readPixels

File: skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc [hunk 14]
```diff
@@ -0,0 +1,424 @@
...
+        ID3D12Resource* resource = texture->resource;
+
+        GrD3DTextureResourceInfo texResInfo = {};
+        texResInfo.fResource.retain(resource);
+        texResInfo.fResourceState = D3D12_RESOURCE_STATE_COMMON;
+        texResInfo.fFormat = DXGI_FORMAT_R8G8B8A8_UNORM;
+        texResInfo.fSampleCount = 1;
+        texResInfo.fLevelCount = 1;
+        GrBackendRenderTarget* renderTarget = new GrBackendRenderTarget(texture->width, texture->height, texResInfo);
+        return reinterpret_cast<jlong>(renderTarget);
```
Comment (elijah-semyonov) [inline skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc:275]: Both `resource` and `readbackBufferResource` allocations can fail inside DirectXOffScreenTexture constructor, we need to check that both are not nullptr or fallback otherwise (delete DirectXOffScreenTexture and return zero).

File: skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc [hunk 15]
```diff
@@ -0,0 +1,424 @@
...
+        return toJavaPointer(d3dDevice);
+    }
+
+    JNIEXPORT jlong JNICALL Java_org_jetbrains_skiko_swing_Direct3DSwingRedrawer_makeDirectXRenderTargetOffScreen(
+            JNIEnv *env, jobject redrawer, jlong texturePtr) {
+        DirectXOffScreenTexture *texture = fromJavaPointer<DirectXOffScreenTexture *>(texturePtr);
+        ID3D12Resource* resource = texture->resource;
+
+        GrD3DTextureResourceInfo texResInfo = {};
+        texResInfo.fResource.retain(resource);
```
Comment (elijah-semyonov) [inline skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc:269]: If `resource` failed to allocated, it's nullptr here. Is it safe?

Comment (Walingar) [reply skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc:269]: If texture is created, it means that resource is allocated
